### DISCUSSION TITLE
spaCy custom NER are not working for new data set.

### DISCUSSION THREAD

**Discussion by arunnaray** [explosion/spaCy#10893] [Help: Coding & Implementations]
I am trying to create few custom NER for my use case :

Training data sample : 

```
[[' webex enable or disable auto join from connected video device advised user to raise a iiq request to get webex access.  hence closing this ticket getting error while joing webex since i got a replacement lap top i have not been able to access webex 954 995 0444 9 30am 6pm thurs mon getting error while joing webex.',
  {'entities': [[1, 62, 'ISSUE'],
    [63, 118, 'RESOLVE_ACTION'],
    [147, 178, 'ISSUE'],
    [193, 212, 'USER_ACTIVITY']]}],
 [' virtual desktop infrastructure vdi performance or latency issues or slow response education vdi user was getting vdi slowness issue while working on it.  i have guide to user kindly logoff your vdi machine and login back after 10 minutes.  user was getting vdi slowness issue while working on it.  vdi slowness.',
  {'entities': [[1, 47, 'ISSUE'],
    [51, 65, 'ISSUE'],
    [69, 96, 'ISSUE'],
    [183, 238, 'RESOLVE_ACTION']]}]]
```
my code is below : 
```
ner = nlp.get_pipe("ner")
for _, annotations in train_data:
    for ent in annotations.get('entities'):
        ner.add_label(ent[2])
disable_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
import random 
from spacy.util import minibatch, compounding
from pathlib import Path

with nlp.disable_pipes(*disable_pipes):
    optimizer =nlp.resume_training()
    
    for iteration in range(100):
        
        random.shuffle(train_data)
        losses = {}
        
        batches = minibatch(train_data,size=compounding(1, 16, 1.001))
        for batch in batches:
            text,annotation = zip(*batch)
            print (text,annotation)
            nlp.update(text,annotation,drop=0.5,losses=losses,sgd=optimizer)
            print("Losses",losses)

output_dir = '/<my model path>/models'
if output_dir is not None:
    output_dir = Path('/my model path>/models')
    if not output_dir.exists():
        output_dir.mkdir()
    nlp.meta['name'] = "new_model"  # rename model
    nlp.to_disk(output_dir)
    print("Saved model to", output_dir)

#Testing model on training data set itself

for text, _ in train_data:
    doc = nlp2(text)
    print ('Entities',[(ent.text,ent.label_) for ent in doc.ents])
    displacy.render(nlp2(doc.text), style="ent", jupyter=True)
```
My models are showing labels for my train_data set, model is not returning anything for any other sample data / text that I am giving.

It just not return anything
<img width="779" alt="Screenshot 2022-06-01 at 8 13 42 PM" src="https://user-images.githubusercontent.com/14104521/171432172-1230c693-776e-4d45-9ba8-d7985e5f966e.png">
.

**Comment by koaning** [comment]
@arunnaray in the future, please don't post screenshots of code. It makes it much harder to copy/paste and it's also impossible to index for search engines. 

Is there a reason why you're writing your own training loop here? As of spaCy v3, the training loop no longer needs to be written and a command-line interface is able to handle all of that on your behalf. You may appreciate [this YouTube video](https://www.youtube.com/watch?v=k77RrmMaKEI) that highlights the transition from spaCy v2 to spaCy v3.

**Comment by polm** [comment]
I notice your entities are very long spans, more like full sentences than a few words. While you can train an NER model on things like that, it won't work very well, and I'm not surprised you can't get useful matches. NER models are really intended for things like names that have clear boundaries and are relatively short.

You might have more luck with the spancat component, which is less particular about span boundaries.

**Reply by polm** [reply]
Also, looking at your examples in more detail, I don't understand your labels - what's the difference between `USER_ACTION` and `ISSUE`? How can you tell when you should have multiple `ISSUE`s in a row vs one long `ISSUE`? Maybe you have clear rules for that, but if it's hard for a human it's going to be hard for the model. You might benefit from going through [the spaCy course](https://course.spacy.io/en/).
